File: angularapp/src/app/services/donation.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from 'src/environments/environment';
import { Donation } from '../models/donation.model';

@Injectable({
  providedIn: 'root'
})
export class DonationService {
  backendUrl=environment.backendUrl;
  httpParam:any;
  constructor(private httpClient:HttpClient) { }

  public getAllDonations():Observable<any>{
    return this.httpClient.get(this.backendUrl+"/donations");
  }

  public getDonationById(donationId:number):Observable<any>{
    return this.httpClient.get(this.backendUrl+"/donations/"+donationId);
  }

  public addDonation(userId:string, charityId:string, donation:Donation):Observable<any>{
    this.httpParam = new HttpParams().set("userId", userId).set("charityId", charityId);
    console.log("service");
    return this.httpClient.post(this.backendUrl+"/donations", donation, {params: this.httpParam});
  }

  public getDonationsByUserId(userId:string):Observable<any>{
    return this.httpClient.get(this.backendUrl+"/donations/user/"+userId);
  }

  public getDonationsByCharityId(charityId:number):Observable<any>{
    return this.httpClient.get(this.backendUrl+"/donations/charity/"+charityId);
  }
  
}

File: springapp/src/main/java/com/examly/springapp/service/DonationService.java
package com.examly.springapp.service;

import java.util.List;

import com.examly.springapp.model.Donation;

public interface DonationService {

    Donation addDonation(Long userId, Long charityId, Donation donation);
    List<Donation> getAllDonations();
    List<Donation> getAllDonationsByUserId(Long userId);
    List<Donation> getAllDonationsByCharityId(Long charityId);
    Donation getDonationById(Long donationId);

}
